designs/2019-plugin-loading-improvement/README.md

```diff
@@ -0,0 +1,285 @@
+
+If the plugins of the personal config conflict with `--plugin` option or something like, then ESLint throws.
+
+#### Implementation
+
+The implementation will be simple.
+
+- Modify [ConfigArrayFactory](https://github.com/eslint/eslint/blob/62623f9f611a3adb79696304760a2fd14be8afbc/lib/cli-engine/config-array-factory.js) to load plugins relatively `basePath`. If `resolvePluginsRelativeTo` constructor option is present, it ignores the `basePath` and loads from the `resolvePluginsRelativeTo`. The `basePath` is:
+  - the directory of `options.filePath` if the `create(data, options)` method is called. It's `cwd` if `options.filePath` is not present.
+  - the directory of `filePath` if the `loadFile(filePath, options)` method is called.
+  - the `directoryPath` if the `loadInDirectory(directoryPath, options)` method is called.
+- Modify [the `mergePlugins` function in ConfigArray](https://github.com/eslint/eslint/blob/62623f9f611a3adb79696304760a2fd14be8afbc/lib/cli-engine/config-array/config-array.js#L163) to throw a clear error if a plugin conflict was found. The conflict is `sourceValue.filePath !== targetValue.filePath`. Both `sourceValue` and `targetValue` are [ConfigDependency](https://github.com/eslint/eslint/blob/62623f9f611a3adb79696304760a2fd14be8afbc/lib/cli-engine/config-array/config-dependency.js) objects, so the error object can have the importer's info as the template data easily.
+
+#### The plugin uniqueness for different lint target files
+
+After this proposal was implemented, ESLint guarantees the plugin uniqueness for each target file individually. Because it's the requirement for we use the rule ID such as `foo/a-rule`. On the other hand, ESLint doesn't guarantee the plugin uniqueness over multiple target files. For example, we can lint `packages/alice/index.js` and `packages/bob/index.js` with each version of the same plugin.
+
+    <root>
+    ├ node_modules/
+    │ └ eslint/
+    ├ packages/
+    │ ├ alice/
+    │ │ ├ node_modules/
+    │ │ │ └ eslint-plugin-foo/  # (A)
+    │ │ ├ index.js
+    │ │ └ .eslintrc.js
+    │ └ bob/
+    │   ├ node_modules/
+    │   │ └ eslint-plugin-foo/  # (B)
+    │   ├ index.js
+    │   └ .eslintrc.js
+    └ .eslintrc.js
+
+When ESLint lints `packages/alice/index.js`, it uses `packages/alice/node_modules/eslint-plugin-foo`. And when ESLint lints `packages/bob/index.js`, it uses `packages/bob/node_modules/eslint-plugin-foo`. Those are not related to each other.
+
+However, only the following APIs require the plugin uniqueness over multiple target files.
+
+- `CLIEngine#getRules()`
+- `metadata.rulesMeta` of formatters
+
+This proposal deprecates those APIs. The following sections describe this.
+
+### ● It replaces the `CLIEngine#getRules()` method
+
+This RFC replaces the `CLIEngine#getRules()` method by `CLIEngine#getRulesForFile(filePath)` because the `CLIEngine#getRules()` method requires the plugin uniqueness over multiple target files.
```

Comment:
This sounds like it might have a major impact on performance. Do we need a separate section about caching plugins for this new method?

Reply:
Thank you for your review.

I think it doesn't have any performance impact. The current logic is [lib/cli-engine/cli-engine.js#L624-L634](https://github.com/eslint/eslint/blob/b41677ae2a143790b19b0e70391a46ec6c8f5de1/lib/cli-engine/cli-engine.js#L624-L634); it merges all rules from all configs to one map. Only if the config files in subdirectories exist, that `lastConfigArrays` may include 2 or more config objects. This is not changed.

When the merging, currently it guarantees every rule object is the same object per rule ID because all plugins are loaded from the same location. But after this proposal, it doesn't guarantee so. Therefore, this section describes adding the deprecation warning.

Reply:
Ah, yes, the current logic is slow.

Reply:
Sorry, my memory is a bit fuzzy on `getRules`, but if I remember it, when I wrote it, it would only load rules once per config (i.e. one time tax for all files that are using the same config). It might've changed since then. It sounds like `getRulesForFile` will load it once per file, which is a lot more expansive. However, I might be wrong in my recollection, or things might've changed since then.

Reply:
Ah, OK. probably I understand your concern. I will update this RFC to mention caches.

The performance impact should be nothing because `CascadingConfigArrayFactory` class caches loaded configs efficiently:

- The cache key is directory paths and each value is the config of the directory.
- The cache value inherits from the parent directory if the directory doesn't have any config file. Therefore, it loads configs only once per config file.